openerp/osv/fields.py

```diff
@@ -861,7 +861,12 @@ def set(self, cr, obj, id, field, values, user=None, context=None):
                     ids2 = act[2] or [0]
                     cr.execute('select id from '+_table+' where '+self._fields_id+'=%s and id <> ALL (%s)', (id,ids2))
                     ids3 = map(lambda x:x[0], cr.fetchall())
-                    obj.write(cr, user, ids3, {self._fields_id:False}, context=context or {})
+                    inverse_field = obj._fields.get(self._fields_id)
+                    if getattr(inverse_field, "ondelete", None) == "cascade":
+                        obj.unlink(cr, user, ids3, context=context)
+                    else:
+                        obj.write(cr, user, ids3, {self._fields_id:False},
+                                  context=context or {})
```

Comment:
About style (sorry, I am picky):
- add a space after the colon `{self._fields_id: False}`
- `context or {}` is pointless, simply use `context`
- don't break the line